File: String/8. String to Integer (atoi)/int.py
class Solution:
    def myAtoi(self, s):
        # Code here
        res = 0
        sign = 1
        index  = 0
        n = len(s)
        
        # ignoring leading whitespace
        while index<n and s[index]==' ':
            index+=1
        
        # store the sign of a number
        while index<n and (s[index]=='-' or s[index]=='+'):
            if s[index] == '-':
                sign = -1
            index+=1
            
        while index<n and '0'<=s[index]<='9':
            res = res*10 + (ord(s[index]) - ord('0'))
            
            if res>2**31-1:
                return (2**31-1) if sign==1 else -2**31

            index+=1
            
        return sign*res

File: String/8. String to Integer (atoi)/int.cpp
class Solution {
public:
    int myAtoi(string s) {
        int n = s.length();
        int sign = 1;
        int index=0;
        long long sum = 0;

        // Handle leading whitespace
        while(index<n && s[index]==' ')
            index++;
        // Handle sign
        if(index<n && (s[index]=='-'||s[index]=='+'))
        {
            sign = (s[index]=='-')?-1:1;
            index++;
        }

        // Process digits and check for overflow
        while(index<n && s[index]>='0' && s[index]<='9')
        {
            int digit = s[index]-'0';

            if(sum>INT_MAX/10 || (sum==INT_MAX/10 && digit>7))
                return (sign==1)?INT_MAX:INT_MIN;

            sum = sum*10+digit;
            index++;
        }

        return sign*sum;
    }
};
